README.md
# Progress Button

[![Pub](https://img.shields.io/pub/v/progress_indicator_button.svg)](https://pub.dartlang.org/packages/progress_indicator_button/) [![Build Status](https://api.cirrus-ci.com/github/PascalAC/progress_button.svg)](https://cirrus-ci.com/github/PascalAC/progress_button) [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=KQ8V6BELJP47G)

A simple button that gives you the possibility to 
transform into a circular one and shows a progress indicator.

![](progress_button_animation.gif)

## How to use

```dart
Container(
    width: 200,
    height: 60,
    child: ProgressButton(
        borderRadius: BorderRadius.all(Radius.circular(8)),
        strokeWidth: 2,
        child: Text(
            "Sample",
            style: TextStyle(
                color: Colors.white,
                fontSize: 24,
            ),
        ),
        onPressed: (AnimationController controller) {
            if (controller.isCompleted) {
                controller.reverse();
            } else {
                controller.forward();
            }
        },
    ),
),
```

Please check the sample folder for a simple implementation.

## Changelog

Please see the [Changelog](CHANGELOG.md) page to know what's recently changed.

## Contributions

Feel free to contribute to this project.

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue](https://github.com/PascalAC/progress_button/issues).  
If you fixed a bug or implemented a new feature, please send a [pull request](https://github.com/PascalAC/progress_button/pulls).

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
## [0.0.3] - 2019-08-02

* A new parameter to customize the progress indicator's stroke width
* Fix a layout bug when the button height is smaller than 60

## [0.0.2] - 2019-02-23

* Remove the need of size parameters by calculating them depending on the parent widget

## [0.0.1] - 2019-02-22

* Initial open source release


sample/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:sample/main.dart';

void main() {
  
}


test/progress_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:progress_indicator_button/button_stagger_animation.dart';
import 'package:progress_indicator_button/progress_button.dart';
import '../example/lib/src/app.dart';

void main() {
  testWidgets('ProgressButton - Start animation on click', (WidgetTester tester) async {
    await tester.pumpWidget(App());

    expect(find.byType(ProgressButton), findsOneWidget);

    await tester.tap(find.byType(RaisedButton));

    final widget = tester.widget<ButtonStaggerAnimation>(find.byType(ButtonStaggerAnimation));
    expect(widget.controller.isAnimating, true);
  });
}


lib/progress_button.dart
library progress_button;

import 'package:flutter/material.dart';
import 'package:progress_indicator_button/button_stagger_animation.dart';

class ProgressButton extends StatefulWidget {
  /// The background color of the button.
  final Color color;
  /// The progress indicator color.
  final Color progressIndicatorColor;
  /// The size of the progress indicator.
  final double progressIndicatorSize;
  /// The border radius while NOT animating.
  final BorderRadius borderRadius;
  /// The duration of the animation.
  final Duration animationDuration;

  /// The stroke width of progress indicator.
  final double strokeWidth;
  /// Function that will be called at the on pressed event.
  ///
  /// This will grant access to its [AnimationController] so
  /// that the animation can be controlled based on the need.
  final Function(AnimationController) onPressed;
  /// The child to display on the button.
  final Widget child;

  ProgressButton({
    @required this.child,
    @required this.onPressed,
    this.color = Colors.blue,
    this.strokeWidth = 2,
    this.progressIndicatorColor = Colors.white,
    this.progressIndicatorSize = 30,
    this.borderRadius = const BorderRadius.all(Radius.circular(0)),
    this.animationDuration = const Duration(milliseconds: 300),
  });

  @override
  _ProgressButtonState createState() => _ProgressButtonState();
}

class _ProgressButtonState extends State<ProgressButton>
    with TickerProviderStateMixin {
  AnimationController _controller;

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      duration: widget.animationDuration,
      vsync: this,
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ButtonStaggerAnimation(
        controller: _controller.view,
        color: widget.color,
        strokeWidth: widget.strokeWidth,
        progressIndicatorColor: widget.progressIndicatorColor,
        progressIndicatorSize: widget.progressIndicatorSize,
        borderRadius: widget.borderRadius,
        onPressed: widget.onPressed,
        child: widget.child,
      ),
    );
  }
}


lib/button_stagger_animation.dart
import 'package:flutter/material.dart';

class ButtonStaggerAnimation extends StatelessWidget {
  // Animation fields
  final AnimationController controller;

  // Display fields
  final Color color;
  final Color progressIndicatorColor;
  final double progressIndicatorSize;
  final BorderRadius borderRadius;
  final double strokeWidth;
  final Function(AnimationController) onPressed;
  final Widget child;

  ButtonStaggerAnimation({
    Key key,
    this.controller,
    this.color,
    this.progressIndicatorColor,
    this.progressIndicatorSize,
    this.borderRadius,
    this.onPressed,
    this.strokeWidth,
    this.child,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: _progressAnimatedBuilder);
  }

  Widget _buttonChild() {
    if (controller.isAnimating) {
      return Container();
    } else if (controller.isCompleted) {
      return OverflowBox(
        maxWidth: progressIndicatorSize,
        maxHeight: progressIndicatorSize,
        child: CircularProgressIndicator(
          strokeWidth: strokeWidth,
          valueColor: AlwaysStoppedAnimation<Color>(progressIndicatorColor),
        ),
      );
    }
    return child;
  }

  AnimatedBuilder _progressAnimatedBuilder(
      BuildContext context, BoxConstraints constraints) {
    var buttonHeight = (constraints.maxHeight != double.infinity)
        ? constraints.maxHeight
        : 60.0;

    var widthAnimation = Tween<double>(
      begin: constraints.maxWidth,
      end: buttonHeight,
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: Curves.easeInOut,
      ),
    );

    var borderRadiusAnimation = Tween<BorderRadius>(
      begin: borderRadius,
      end: BorderRadius.all(Radius.circular(buttonHeight / 2.0)),
    ).animate(CurvedAnimation(
      parent: controller,
      curve: Curves.easeInOut,
    ));

    return AnimatedBuilder(
      animation: controller,
      builder: (context, child) {
        return SizedBox(
          height: buttonHeight,
          width: widthAnimation.value,
          child: RaisedButton(
            shape: RoundedRectangleBorder(
              borderRadius: borderRadiusAnimation.value,
            ),
            color: color,
            child: _buttonChild(),
            onPressed: () {
              this.onPressed(controller);
            },
          ),
        );
      },
    );
  }
}


example/lib/src/app.dart
import 'package:flutter/material.dart';
import 'package:progress_indicator_button/progress_button.dart';

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Progress Button Sample",
      home: Scaffold(
        appBar: AppBar(
          title: Text("Sample"),
        ),
        body: Center(
          child: Container(
            width: 200,
            height: 60,
            child: ProgressButton(
              borderRadius: BorderRadius.all(Radius.circular(8)),
              strokeWidth: 2,
              child: Text(
                "Sample",
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                ),
              ),
              onPressed: (AnimationController controller) {
                if (controller.isCompleted) {
                  controller.reverse();
                } else {
                  controller.forward();
                }
              },
            ),
          ),
        ),
      ),
    );
  }
}
